File: arrays/first-missing-integer.cpp
int Solution::firstMissingPositive(vector<int> &A) {
    int n = A.size();

    for(int i = 0; i < n; i++){
        while(A[i] > 0 && A[i] <= n && A[i] != i+1 && A[A[i]-1] != A[i]){
          swap(A[A[i]-1], A[i]);
        }
    }

    for(int i = 0; i < n; i++) if(A[i] != i+1) return i+1;

    return n+1;
}

File: arrays/first-missing-integer.py
class Solution:
    def firstMissingPositive(self, A):
        i = 0
        while i < len(A):
            if 0 < A[i] <= len(A):
                x = A[i]
                A[i] = A[x-1]
                A[x-1] = x
                if A[i] != x: continue
            i += 1

        for i in range(len(A)):
            if A[i] != i+1: return i+1

        return len(A)+1

